Add Felix calc-graph and dataplane design docs (#12977)
projectcalico/calico: felix/design/calc-graph.md

datastore updates from the syncer and emits protobuf messages
describing the desired state of *this node's* dataplane in
Calico-internal terms (local WEPs, resolved policy rules,
fully-expanded IP sets, routes, VTEPs). Its output is a function of
current datastore state, computed incrementally so it can keep up
with churn.

It was built for **100k+ endpoints, ~2k+ policies, ~1000+
endpoint-updates/second**, with the dataplane applying the result
sub-second. That drives three properties:

- **Incremental, minimal-delta output** so the dataplane applies the
  smallest possible change set with low latency.
- **Back-pressure.** If the graph or dataplane can't keep up,
  upstream layers block; blocking lets them coalesce queued work
  (see the `EventSequencer`) instead of growing an unbounded backlog.
- **Local filtering.** Only ~100 local WEPs need policy computed, not
  the 100k in the cluster, so the graph reduces cluster state to the
  local subset early (the `localEndpointDispatcher`, below).

Centralising this keeps every dataplane simpler. The graph is the
single place that handles:

1. **Cross-resource inconsistency** from eventual consistency — a WEP

>    naming a not-yet-known profile, a rule selecting an unknown label.

Reviewer: ```suggestion
   naming a not-yet-known profile, a rule selecting a label that doesn't yet match any endpoints, or whose matches can change later.
```

Reviewer: (because there's no such thing as an "unknown" label in the sense of a missing resource for it)

2. **Transient duplicate state** — e.g. two WEPs briefly sharing an
   IP after add/delete reordering; the graph resolves the conflict so
   no dataplane has to.

> 3. **Local relevance filtering** (problem above).

Reviewer: what "problem above" ?

Reason about each event on its own, against your prior knowledge of
*that one resource* — not about the sequence as a whole.

- **`nil` value** = treat as "doesn't exist" (deleted, or failed
  validation and treated as absent). Validation is upstream in
  `calc/validation_filter.go` (`ValidationFilter` nils out invalid
  values rather than altering them), so a non-nil value has already
  passed schema/semantic validation.
- **Non-nil value** = the resource's current state; compare against
  what you held and reconcile.

The **only** real guarantee is eventual convergence to the latest
value (supported datastores make writes durable, so connectivity
permitting you will eventually see it). Everything else is fair
game. For a resource that truly went `Created → A → B → C`, a node
might observe any of:

- `A → B → C`
- `A → C` — `B` coalesced away
- `C` — `A` and `B` coalesced away
- `A → C → B → C` — reached `C`, then a resync hit a stale replica
  (back to `B`), then caught up
- `C → Deleted → A → B → C` — resync hit a replica so stale the
  resource didn't exist there yet, giving a **spurious `Deleted`**

So a node must not assume ordering, monotonic versions, that it sees

> every transition, or even that a `Deleted` is real and final —

Reviewer: I'm not sure what this line is advising in practice.  How could code be wrong by making an incorrect "assumption that a `Deleted` is real and final"?  I'm pretty sure this isn't saying "ignore Deletes" - so what **is** it saying?

Reviewer: (I believe there are already many cases where the same resource can be Deleted and then re-Created.)

"compare the event against my state and reconcile" handles all of
these; "assume values only move forward" does not.

### The update-type side channel

Each event also carries `Update.UpdateType` (`api.UpdateType`:
`UpdateTypeKVNew`/`KVUpdated`/`KVDeleted`/`KVUnknown` in
`libcalico-go/lib/backend/api`). Original purpose was stats
without retaining objects (see `calc/stats_collector.go`).

> Treat it as advisory, and note that

Reviewer: Does "Treat it as advisory" mean "Usually ignore this UpdateType".

`UpdateTypeKVNew/Updated` can carry `nil` due to failed
validation.

### Review notes

- A node that decides *existence* from the update type rather than

>   nil/non-nil is suspect; stats-style bookkeeping is the legitimate

Reviewer: ```suggestion
  nil/non-nil is suspect; stats-style bookkeeping is the only legitimate
```

Reviewer: (if that is correct.  If not, please state the other legitimate uses)

| `L3RouteResolver` (`l3_route_resolver.go`) | Indexes IPAM blocks, IP pools and node metadata into longest-prefix-match routes over Calico-known IP space (CIDR + pool type/metadata + is-host + owning host for workloads); consumed by the BPF and VXLAN dataplanes |
| `VXLANResolver` (`vxlan_resolver.go`) | Resolves node IP/config into a VTEP per host (`proto.VXLANTunnelEndpointUpdate`/`Remove`); the dataplane only programs VXLAN routes once the VTEP is ready |
| `EncapsulationResolver` (`encapsulation_resolver.go`) | Derives encap mode from IP-pool config (restarts Felix if it changed) |
| `IstioCalculator` (`istio_calculator.go`) | Marks local WEPs that are in the Istio ambient mesh so downstream can apply mesh networking |
| `EventSequencer` (`event_sequencer.go`) | Output stage: buffers, coalesces, flushes in dependency order |

### Review notes

- A PR adding a consumer of "A matches B" events should state where it
  sits relative to the producers of A and B, and handle the teardown
  order. Any buffer must rely only on same-loop delivery.
- A PR reordering node registration must justify it against the
  matches-after-members pattern; most reorderings are inert, but the

>   `LiveMigrationCalculator`-before-ARC one is silently load-bearing.

Reviewer: ```suggestion
  `LiveMigrationCalculator`-before-ARC one is load-bearing.
```

Reviewer: (it's not "silent" if the code comments it - which is stated a few paragraphs above)

### Review notes

- Preserve exact add/remove key symmetry; audit both directions.
- Keep the shared-IP/overlapping-membership cases working, and extend
  the adversarial index tests rather than only adding happy-path
  coverage.
- Reject creation of additional all-endpoint indexes.  Prefer refactoring
  to share existing indexes.
- Consider occupancy: avoid adding fields to per-endpoint structs,
  every field costs many MB at scale. Custom datastructures can
  be justified with benchmarks.

## The EventSequencer (output stage)

`EventSequencer` (`calc/event_sequencer.go`) is the output boundary.
It buffers updates in `pending*` maps/sets and emits them on
`Flush()`, coalescing repeated changes to the same object in between.

### Coalescing is back-pressure

Coalescing is half the back-pressure mechanism. When the dataplane
stalls on a big update while the datastore churns, coalescing
**bounds memory** (at most one pending entry per resource, so ~the
cluster's resource count) and **bounds the update size** handed to
the dataplane. Worst case degrades gracefully: the dataplane runs a
catch-up loop, each pass absorbing the previous pass's churn, rather

> than growing an unbounded queue.

Reviewer: I'm not sure I follow this.  I wasn't familiar with how the calc graph experiences back pressure, but I looked that up and now see that it's the `select` in `func (acg *AsyncCalcGraph) onEvent` - i.e. `onEvent` blocks if any of `acg.outputChannels` are full.  With that in mind, what exactly do the following parts of the above mean:

> at most one pending entry per resource

Does "resource" mean one of the `proto.` message types?  If not, what does it mean?  What does "pending" mean - i.e. whereabouts in the overall flow from the datastore to the dataplane?

> bounds the update size

Is that number of proto messages, or the size of each of those messages; or a bit of both?

remove before theirs; if it has no dependencies, join the loose tail.

### The missing-resource tension

When a referent is genuinely missing (not just late), there are
three sanctioned strategies. Which fits depends on the cost of
handling it in the dataplane versus buffering in the graph, and on
what makes sense for the resource type:

- **(a) Synthesize a safe stand-in.** Felix does this for profile
  rules: a missing profile resolves to a fail-safe deny-all rule set
  (`DummyDropRules` in `calc/active_rules_calculator.go`). Some
  resource types have no meaningful stand-in.
- **(b) Pass the inconsistency through to the dataplane** as a signal
  and let it fail closed on its own terms. The right choice when the
  dataplane must act anyway — e.g. a WEP gains a security-critical
  field, so a missing dependency means *that endpoint* must fail
  closed.
- **(c) Buffer the dependent until the dependency arrives.**
  Requires sending a remove of any already-sent copy of the dependent
  in order to maintain the calc-graph's "no hysteresis" invariant.
  Never buffer endpoints/policies/profiles; these are security-critical

>   and part of the core feature set.

Reviewer: Is this last sentence part of bullet (c), or a more global instruction?  It reads more like the latter and so should be reformatted as such.

The invariant across all three: a missing dependency must never
silently leave an endpoint or policy *open* — it must fail closed.

### Review notes

- A new message type must document its place in the flush order (and
  why) and carry FV coverage — wrong-phase bugs only bite under
  specific orderings, which FV expansion is built to catch.
- Handling a missing referent by buffering an endpoint or policy is
  almost always wrong; prefer (a) or (b).

> - Removing or weakening a coalescing path must argue it doesn't break

Reviewer: For me, this point is sufficiently obscure that it would help to point to an example.

companion runs (unless `DISABLE_TEST_EXPANSION=true`):

- **`reverseKVOrder`** — reverse KV order within each state (output
  mustn't depend on intra-state delivery order).
- **`reverseStateOrder`** — reverse the states (tests build-up and
  teardown).
- **`insertEmpties`** — empty state between each pair (create, tear
  down, recreate).
- **`splitStates`** — each state standalone from empty
  (self-consistency in isolation).
- **`squashStates`** — collapse the whole sequence (incl. deletions)
  into one state via `KVDeltas` (same end state in one step as
  incrementally).

**Blind spot:** a symmetric sequence like `[{A}, {A,B}, {A}]` is its
own reverse, so it only ever tests "A before B" — never the
`{A,B} → {B}` transition (A deleted while B still references it). Add
an explicit asymmetric sequence when the teardown-with-live-referrer
case matters (it usually does for indexes/refcounts).

### Review notes

- A calc-graph (or labelindex) change without an FV state is the
  exception and needs justification; per-node unit tests aren't a
  substitute — funnel coverage through the FV framework.
- Check new states aren't accidentally symmetric (the

>   `[{A},{A,B},{A}]` trap); if so, add an asymmetric sequence.

Reviewer: This is a sequence, not a state, right?

In any case, it doesn't seem to me that a symmetric sequence would be necessarily wrong.  I think the important point here was already made under "Blind spot", and the bullet here can be deleted.

## Common failure modes

1. **Buffering "good" output across an inconsistency** — violates
   depend-only-on-current-state. Sequence output nicely, but its
   content must be a function of current state; if a resource is
   broken, emit a fail-closed stand-in or pass-through signal rather
   than withholding a security-critical update.
2. **Buffering security-critical resources** (endpoints, policies) —
   fail closed instead.
3. **Recomputing work that scales with cluster-wide counts on a
   high-churn path** (the historical 20,000-sorts bug).
4. **Leaky/asymmetric refcounting** — add and remove keyed
   differently.
5. **Skipping the FV suite** — so the orderings and teardown paths

>    the expanders would catch go untested.

Reviewer: These points are slightly confusing, in that sometimes the part after the em-dash is "explanation about why this is wrong", and sometimes it is "this is what to do instead".  The "add and remove keyed differently" one risks being read either way, with opposite meanings.

   is eventually consistent, so a resource can reference another
   that the graph hasn't heard about yet (a WEP naming a profile
   that doesn't exist; a rule selecting a not-yet-known label).
2. **Transient duplicate state.** Add/delete reordering can
   briefly present two WEPs with the same IP. The graph detects the
   conflict and decides how to resolve it, so the dataplane never
   has to.
3. **Local relevance filtering.** Reducing cluster state to this
   node's dataplane state (problem (c) above).
4. **Dataplane-friendly output ordering.** Dependencies are always
   emitted *before* the resources that depend on them, and are only
   removed once nothing depends on them any more. This is the
   `EventSequencer`'s job (below) and is the contract the dataplane
   relies on.

### Review notes for this section

- A PR that proposes to "recompute" some derived state wholesale on
  each update is fighting the architecture. The default is
  incremental, edge-triggered processing. Wholesale recomputation
  is acceptable only where the recomputed set is provably bounded
  by *local* state (e.g. re-sorting one local endpoint's policy

>   list) and never by cluster-wide state.

Author: Recomputing wholesale is actually valid for anything small or low churn, not just the example above.  For example, it's perfectly fine to recompute felix configuration from scratch when the handful of felix config resources change.

Full recomputation from cached state (with dedupe on output) is always the simplest option, so it should be used wherever it's the right trade-off.

Author: Good point — reworded the review note in 2d783ae: wholesale recomputation from cached state (with output dedup) is now described as the *simplest* option and fine for small/low-churn state (Felix config called out as an example). The anti-pattern is narrowed to cluster-wide-scaled work on high-churn paths.

- A PR that adds per-update work proportional to the *cluster-wide*
  resource count (not the local count) is a scaling regression and
  needs an explicit justification or a redesign. This is the

>   calc-graph analogue of the BPF "fast-path cost" rule.

Author: ```suggestion
  needs an explicit justification or a redesign.
```
Not sure we need to ref the BPF quote.

Author: Done in 2d783ae — dropped the BPF analogue sentence.

endpoints **in the same calc-graph loop** as the match event, so
the buffer you need is small and short-lived. And remember the
symmetric teardown: if you heard A and B before "A matches B", then
on deletion you will hear A and B removed *first*, then "A no longer
matches B".

### The principal nodes

| Node (file) | Role |
|---|---|
| `ValidationFilter` (`validation_filter.go`) | Nils out invalid resources (treat-as-missing) before they reach the graph |
| `AllUpdDispatcher` / `localEndpointDispatcher` (`dispatcher` pkg) + `endpointHostnameFilter` (`calc` pkg) | Type-based fan-out; the filter does the local-endpoint reduction |
| `LiveMigrationCalculator` (`live_migration_calculator.go`) | Adjusts endpoint state for KubeVirt-style live migration |
| `ActiveRulesCalculator` (`active_rules_calculator.go`) | Tracks which policies/profiles are active given local endpoint labels |
| `RuleScanner` (`rule_scanner.go`) | Extracts selector/named-port references from active rules; drives the label index |
| `PolicyResolver` / `PolicySorter` (`policy_resolver.go`, `policy_sorter.go`) | Computes the ordered per-endpoint policy list (tiers, order) |

> | `L3RouteResolver` (`l3_route_resolver.go`) | Computes routes from IP pools, WEPs, host IPs |

Author: ```suggestion
| `L3RouteResolver` (`l3_route_resolver.go`) | Computes a generalized route map of Calico-known IP space from IP pools, WEPs, host IPs. Generalized in the sense that some entries are just "we know this useful information about this CIDR" rather than actual IP routes. |
```

Author: Applied your suggestion verbatim in 2d783ae.

1. Identify its dependencies (what must already be in the dataplane
   before this message is safe to apply) and its dependents (what
   must still be present when this message is removed).
2. Place its **add** after its dependencies' adds; place its
   **remove** before its dependencies' removes (i.e. mirror it).
3. If it has no dependency relationship, it can join the loosely-
   ordered tail.

### The missing-resource tension

Strict ordering collides with referential inconsistency: what does
the graph emit when a referent is genuinely missing (not just late)?
There are three sanctioned strategies, in rough order of preference:

- **(a) Synthesize a safe stand-in.** Best when a safe default
  exists. Felix does this for profile rules: a missing profile is
  resolved to a fail-safe **deny-all** rule set (the `DummyDropRules`
  in `calc/active_rules_calculator.go`), so policy still resolves and
  fails closed.
- **(b) Buffer the dependent until the dependency arrives.** Usually
  the **wrong** choice. You must **not** buffer endpoints or
  policies — they are security-critical, most dependency chains
  start at endpoints, and delaying them risks leaving traffic

>   unsecured. Reserve buffering for non-security-critical leaves.

Author: ```suggestion
  start at endpoints, and delaying them risks leaving endpoints with stale policy/configuration, which could be a security hole. Reserve buffering for non-security-critical leaves.
```

Author: Applied your suggestion in 2d783ae.

- **(c) Make an explicit exception and handle it in the dataplane.**
  Pass the inconsistency through to the dataplane as a signal rather
  than resolving it in the graph, letting the dataplane fail closed
  on its own terms. Use this sparingly, and only when neither a safe

>   stand-in (a) nor buffering (b) fits.

Author: Not sure the advice s quite right, the trade off depends on how much work it is to handle in the dataplanes vs buffering.  If we have to do something in the dataplane anyway (for example, this is a WEP with a new security critical field, so a missing dependency means we need the endpoint to fail closed, buffering is not right and synthesizing a dummy resource might not make any sense for the particular type of resource.

Author: Reframed in 2d783ae: (a)/(b)/(c) are now presented as a trade-off (dataplane work vs buffering, and what fits the resource type) rather than a fixed preference order. Noted a dummy stand-in doesn't always make sense, and that (c) is the right call when the dataplane must act anyway (your new-security-critical-field example). The hard invariant is restated as 'fail closed; which mechanism is the judgement call'.

<!--
Copyright (c) 2026 Tigera, Inc. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

# Felix calculation graph — Design

This is the design doc for Felix's **calculation graph** (the

> "brain"): the code under [`felix/calc/`](../calc/) plus the

Reviewer: Does saying `the "brain"` really add anything?  I think it's better if one thing (the calc graph) has one name.

Reviewer: (If you agree, there are a few other occurrences in this PR.)

Author: _Claude:_ Agreed — dropped the "brain" nickname in the title, intro and the instruction file (and there are no other occurrences left). ed17d72.

   node's dataplane state (problem (c) above).
4. **Dataplane-friendly output ordering.** Dependencies are always
   emitted *before* the resources that depend on them, and are only
   removed once nothing depends on them any more. This is the
   `EventSequencer`'s job (below) and is the contract the dataplane
   relies on.

### Review notes for this section

- Wholesale recomputation from cached state (recompute the output
  from scratch, then dedup no-op changes downstream) is the
  *simplest* option and is perfectly fine wherever the recomputed
  set is small or low-churn — re-sorting one local endpoint's policy
  list, or recomputing Felix configuration from the handful of
  config resources, are both legitimate. Use it wherever it's the
  right trade-off; reach for fine-grained incremental updates only
  when full recomputation would be too expensive.
- The actual anti-pattern is per-update work whose cost scales with
  the *cluster-wide* resource count on a *high-churn* path — e.g.
  touching all 100k endpoints every time any one of them changes.
  That is a scaling regression and needs an explicit justification
  or a redesign.

## The node model and the node contract

> Calculation nodes are **plain function objects wired into a call

Reviewer: Don't know what "plain function objects" means.  "Synchronous functions" perhaps?  But honestly I'm having to guess what the point here is.

Author: I've seen new folks read "calc _graph_" and picture a graph of goroutines and channels, was trying to capture the fact that it's a DAG of nodes, but the links between them are simple function calls, and hence no locking or concurrency is implied.

graph**, not goroutines. An update entering the graph propagates
synchronously through the nodes via direct method calls
(`OnUpdate`, registered callbacks) and the call returns when
propagation is complete. There is no concurrency between nodes, so
nodes need no locking. (Hidden internal parallelism inside a single
node is *conceivable* if it earned its keep, but the bar is high
and it must remain an invisible implementation detail behind the
synchronous API.)

Every node must honour the following contract.

### Handle add, update and delete — in any order

The syncer can deliver KVs in any order and can coalesce them
(below). A node must react correctly to a create, an update, or a

> delete for any resource it tracks, arriving at any time.

Reviewer: But for a given single resource I believe we can assume seeing updates in the right order, right?  E.g. something is created before it is deleted.

Author: If you mean the KV type field then yes, but it's quite subtle.  

If a resource goes through states `Created -> A -> B -> C`, the only real guarantee is that, connectivity permitting, you'll eventually see C.  You might see

- `A -> B -> C`
- `A -> C` (common scenario: B skipped due to colaescing)
- `C` (common scenario: A and B skipped due to coalescing)
- `A -> C -> B -> C`  (rare but possible scenario: got to C then resynced against a stale replica, then catch up to C)
- `C -> Deleted -> A -> B -> C` (same but very stale replica that hasn't even reached A yet)

### Handle inconsistency, and the consistent→inconsistent transition

A node must cope with referential inconsistency — a WEP that names a
non-existent profile, a rule selecting an unknown endpoint — and
produce *some* well-defined output for it.

The subtle, load-bearing half of this: a node must handle the
transition **from consistent to inconsistent** and end up in
exactly the same state as if the inconsistency had been present
from start of day. If resource B disappears while A still
references it, the node must reconcile to the "B is missing"
output — it must **not** cling to the last-known-good value of B.

### Be memoryless

> This is the single most important node invariant. **A node's output

Reviewer: Delete first sentence.  It adds nothing IMO.  Assuming this was mostly Claude-written, I think Claude itself needs a guideline: be less verbose when possible.

Author: _Claude:_ Deleted that sentence, and took the wider point: did a full concision pass on both docs (calc-graph.md down ~25%), cutting filler and editorialising. ed17d72.

  time all the logs and context that would explain it are gone.

Buffering for **work-avoidance** is a different thing and is
allowed (see no-op suppression below, and the `EventSequencer`):
that kind of buffer changes *when* and *how efficiently* output is
produced, never *what* the output is for a given datastore state.

### Suppress no-op churn where it matters

A node should suppress emitting a downstream update when nothing it
cares about actually changed — but only where it matters, i.e.
where the downstream would otherwise do significant redundant work.
The canonical example (since fixed): WEP↔policy match changes used
to re-trigger sorting the policy list for the affected endpoint. At
start of day with 100 local WEPs and 200 policies each, the naive
version sorted 200 × 100 = 20,000 times instead of once per WEP =
100 times. Suppression here was worth it; suppression on a leaf
that does no downstream work is not worth the complexity.

### Edge-triggered indexing must pair add/remove with identical keys

Where a node maintains reference counts or membership indexes (the
label indexes are the prime example, below), the add and remove

> operations are **edge-triggered** and must be keyed **identically**.

Reviewer: What does "edge-triggered" mean?

Author: _Claude:_ Dropped the jargon — it now just says index **add/remove must be balanced and identically keyed** (one remove per add, same key, or it leaks). ed17d72.

  synchronous model doesn't fit.

## The upstream (syncer) contract

The syncer delivers an **eventually-consistent sequence of KV
events**. A node should reason about each event in isolation,
against its own previous knowledge of that one resource — not about
the sequence as a whole.

- **`nil` value** means the resource does not exist: either it was
  deleted, or it failed validation and must be treated as
  non-existent. (Validation lives upstream in
  `calc/validation_filter.go` — `ValidationFilter` nils out invalid
  values rather than altering them — so by the time a non-nil value
  reaches a node it has passed schema/semantic validation.)
- **Non-nil value** means "this is the resource's current state;
  update yourself to match." The node compares against whatever it
  held before and reconciles.
- **Coalescing.** The syncer may collapse a run of updates to the
  same resource — `update → update → delete → update → update` can
  arrive as just the final `update`. A node must never assume it
  sees every intermediate state.
- **Reversion is possible.** A datastore resync that lands on a
  node with a stale cache can legitimately move a resource

>   *backwards* to an earlier version. "Compare current event against

Reviewer: But, to be clear, there will then be following update(s) to reach the actual latest version, right?

Author: Correct

it. A second dispatcher, the `localEndpointDispatcher`, carries the
**locally-filtered** endpoint stream — an `endpointHostnameFilter`
(in the `calc` package) is registered first on it and forwards only
endpoints hosted on this node, which is where the cluster→local
reduction (problem (c)) happens. (The dispatcher itself is
hostname-agnostic; the filter does the reduction.)

Most wiring order is irrelevant because nodes are independent. The
order that *does* matter follows one pattern: **a node that consumes
"A matches B" events generally wants to have already heard about A
and B individually before it hears that they match.** Otherwise it
has to buffer the match until A and B turn up.

The one concrete load-bearing ordering in `NewCalculationGraph` is
about **handler registration order on the same dispatcher**, not
about the order the dispatchers are created (`dispatcher.Register`
appends per key-type and `OnUpdate` then iterates handlers in
registration order):

- The `LiveMigrationCalculator` (`calc/live_migration_calculator.go`)
  has its `OnUpdate` registered on the `localEndpointDispatcher`
  **before** the `ActiveRulesCalculator` registers with that same

>   dispatcher, so the ARC sees migration-adjusted endpoint state.

Reviewer: I couldn't understand "so the ARC sees migration-adjusted endpoint state" here.  The actual comment says the right thing: "This ensures the LMC sees WEP updates first, so its wepData is already populated when the ARC fires computed selector match callbacks."

Author: _Claude:_ Replaced my paraphrase with the accurate rationale from the code comment: the LMC's OnUpdate is registered before the ARC's so the LMC sees WEP updates first and its `wepData` is populated when the ARC fires computed selector-match callbacks. ed17d72.

  This is the only such constraint the code calls out in a comment.

If you genuinely must consume a "matches" event before the
endpoints it references, you may rely on hearing about the
endpoints **in the same calc-graph loop** as the match event, so
the buffer you need is small and short-lived. And remember the
symmetric teardown: if you heard A and B before "A matches B", then
on deletion you will hear A and B removed *first*, then "A no longer
matches B".

### The principal nodes

| Node (file) | Role |
|---|---|
| `ValidationFilter` (`validation_filter.go`) | Nils out invalid resources (treat-as-missing) before they reach the graph |
| `AllUpdDispatcher` / `localEndpointDispatcher` (`dispatcher` pkg) + `endpointHostnameFilter` (`calc` pkg) | Type-based fan-out; the filter does the local-endpoint reduction |

> | `LiveMigrationCalculator` (`live_migration_calculator.go`) | Adjusts endpoint state for KubeVirt-style live migration |

Reviewer: To me "Adjusts endpoint state" means "modifying the provided model.WorkloadEndpoint" - which is absolutely not true.  What LMC actually does is correlate between input WEPs and LiveMigration resources and emit proto.WEPs to reflect that.  Oh, and also it's for OpenStack as well as KubeVirt.

Reviewer: It worries me that there's a bit too much "make something up in roughly the right ballpark" in this PR, when we have existing code comments that are correct, e.g.:
```
// LiveMigrationCalculator tracks local workload endpoints and LiveMigration resources and
// correlates between them, in order to correctly set the live_migration_role field in
// proto.WorkloadEndpoints that are sent to the dataplane.
```

Author: _Claude:_ Fixed. The row now tracks the godoc: the LMC **correlates local WEPs with LiveMigration resources to set the `live_migration_role` field on emitted `proto.WorkloadEndpoint`s (OpenStack and KubeVirt)** — no mutation of endpoint state. On your broader concern (make-something-up vs. the real comments): I regrounded every node row in that table on its actual godoc, and pulled the real comments before rewriting rather than paraphrasing from memory. ed17d72.

| `VXLANResolver` (`vxlan_resolver.go`) | Computes VTEP entries |
| `EncapsulationResolver` (`encapsulation_resolver.go`) | Derives encap mode from IP-pool config |
| `IstioCalculator` (`istio_calculator.go`) | Marks WEPs in the Istio ambient mesh |
| `EventSequencer` (`event_sequencer.go`) | Output stage: buffers, coalesces, flushes in dependency order |

### Review notes for this section

- A PR that adds a consumer of "A matches B"-style events should
  state where it sits relative to the producers of A and B, and
  confirm it handles the teardown order (A/B removed before the
  un-match). If it buffers, the buffer must rely only on
  same-loop delivery, not on cross-loop retention.
- A PR that reorders node registration must justify it against the
  "matches after members" pattern; most reorderings are inert, but
  the one that isn't — `LiveMigrationCalculator`'s `OnUpdate`
  registered on `localEndpointDispatcher` before the ARC's — is
  silently load-bearing.

## Label indexes and refcounting

[`felix/labelindex/`](../labelindex/) holds the graph's most

> intricate machinery: `InheritIndex` (label inheritance from

Reviewer: This editorialisation is a tad annoying...  I hope I'm criticising Claude rather than you @fasaxc !

In general I would prefer:
- "show, don't tell" - i.e. just list what the thing does.  Cut the "this next part is amazing, you're going to love it" kind of thing.
- Across the board, more conciseness.

Author: _Claude:_ Fair — cut the editorialising and switched to show-don't-tell across both docs, and regrounded descriptions on the code comments rather than characterising them. ed17d72.
